File: services/kyc-service/src/main/java/dev/mathalama/kycservice/application/mapper/KycMapper.java
package dev.mathalama.kycservice.application.mapper;

import dev.mathalama.kycservice.application.dto.response.KycApplicationResponse;
import dev.mathalama.kycservice.domain.model.KycApplication;

public final class KycMapper {

    private KycMapper() {
    }

    public static KycApplicationResponse toResponse(KycApplication app) {
        if (app == null) {
            return null;
        }
        return KycApplicationResponse.builder()
                .id(app.getId())
                .userId(app.getUserId())
                .status(app.getStatus())
                .documentType(app.getDocumentType())
                .livenessScore(app.getLivenessScore())
                .faceMatchScore(app.getFaceMatchScore())
                .mrzValid(app.getMrzValid())
                .firstName(app.getExtractedFirstName())
                .lastName(app.getExtractedLastName())
                .documentNumber(app.getExtractedDocumentNumber())
                .dateOfBirth(app.getExtractedDateOfBirth())
                .expiryDate(app.getExtractedExpiryDate())
                .nationality(app.getExtractedNationality())
                .rejectionReason(app.getRejectionReason())
                .createdAt(app.getCreatedAt())
                .updatedAt(app.getUpdatedAt())
                .build();
    }
}

File: app/lib/data/mappers/kyc_mapper.dart
import '../../domain/entities/document_type.dart';
import '../../domain/entities/kyc_application_entity.dart';
import '../../domain/entities/kyc_status.dart';
import '../models/kyc_response_model.dart';

extension KycResponseModelMapper on KycResponseModel {
  KycApplicationEntity toDomain() {
    return KycApplicationEntity(
      id: id,
      userId: userId,
      status: KycStatus.fromString(status),
      documentType: DocumentType.fromString(documentType),
      livenessScore: livenessScore,
      faceMatchScore: faceMatchScore,
      mrzValid: mrzValid,
      firstName: firstName,
      lastName: lastName,
      documentNumber: documentNumber,
      dateOfBirth: dateOfBirth,
      expiryDate: expiryDate,
      nationality: nationality,
      rejectionReason: rejectionReason,
      createdAt: createdAt != null ? DateTime.tryParse(createdAt!) : null,
      updatedAt: updatedAt != null ? DateTime.tryParse(updatedAt!) : null,
    );
  }
}
